bson/primitive/objectid.go

```diff
@@ -94,6 +94,18 @@ func IsValidObjectID(s string) bool {
 	return err == nil
 }
 
+// MarshalText returns the ObjectID as UTF-8-encoded text. Implementing this allows us to use ObjectID
+// as a map key when marshalling JSON. See https://pkg.go.dev/encoding#TextMarshaler
+func (id ObjectID) MarshalText() ([]byte, error) {
+	return []byte(id.Hex()), nil
+}
+
+// UnmarshalText populates the byte slice with the ObjectID. Implementing this allows us to use ObjectID
+// as a map key when unmarshalling JSON. See https://pkg.go.dev/encoding#TextUnmarshaler
+func (id ObjectID) UnmarshalText(b []byte) error {
+	return id.UnmarshalJSON(b)
```

Comment:
`ObjectID.UnmarshalJSON()` accepts multiple input formats:

1. the string "null"
2. a 12 byte raw representation
3. an empty string (which unmarshals to a NilObjectID)
4. a 24 character string hex representation
5. a representation of the [canonical extended JSON format](https://github.com/mongodb/specifications/blob/df6be82f865e9b72444488fd62ae1eb5fca18569/source/extended-json.rst#conversion-table): `{"$oid": <ObjectId bytes as 24-character, big-endian hex string>}`

@Luberry would a preferable alternative be to use `ObjectIDFromHex`? It only accepts a 24 character hex string. That may better align with the intent of the `UnmarshalText` interface.

Reply:
I thought of that, I figured this would make sure we we handle extracting them the same as is done with json. So that if someone had a file expecting any other possible formats they could handle it. I'm not too attached either way. And ultimately it's up to you all.

Reply:
I agree with @kevinAlbs's suggestion to limit the implementation to strictly unmarshal from the 24-char string hex representation (i.e. use `ObjectIDFromHex()`). I think that avoids having to support unexpected formats without reducing the usefulness of the feature.

Reply:
alright, ill implement that.